File: clankerprof/proto.py
from __future__ import annotations

import gzip
import zlib

from dataclasses import dataclass, field, replace

from clankerprof.model import (
    Function,
    Location,
    Profile,
    Sample,
    ValueType,
    select_primary_value_index,
)


class PprofDecodeError(ValueError):
    """Raised when profile bytes are not a supported pprof protobuf profile."""


_U64_MASK = (1 << 64) - 1
_INT64_SIGN_BIT = 1 << 63


def _to_signed64(value: int) -> int:
    """Reinterpret an unsigned varint as a two's-complement int64."""
    value &= _U64_MASK
    return value - (1 << 64) if value & _INT64_SIGN_BIT else value


@dataclass(slots=True)
class _Reader:
    data: bytes
    pos: int = 0

    def eof(self) -> bool:
        return self.pos >= len(self.data)

    def read_byte(self) -> int:
        if self.pos >= len(self.data):
            raise PprofDecodeError("Unexpected end of protobuf stream.")
        value = self.data[self.pos]
        self.pos += 1
        return value

    def read_varint(self) -> int:
        shift = 0
        result = 0
        while True:
            byte = self.read_byte()
            result |= (byte & 0x7F) << shift
            if byte < 0x80:
                return result
            shift += 7
            if shift > 70:
                raise PprofDecodeError("Invalid protobuf varint.")

    def read_key(self) -> tuple[int, int]:
        key = self.read_varint()
        return key >> 3, key & 0x07

    def read_length_delimited(self) -> bytes:
        length = self.read_varint()
        end = self.pos + length
        if end > len(self.data):
            raise PprofDecodeError("Length-delimited field extends beyond stream.")
        payload = self.data[self.pos : end]
        self.pos = end
        return payload

    def skip(self, wire_type: int) -> None:
        if wire_type == 0:
            self.read_varint()
            return
        if wire_type == 1:
            self.pos += 8
            return
        if wire_type == 2:
            self.read_length_delimited()
            return
        if wire_type == 5:
            self.pos += 4
            return
        raise PprofDecodeError(f"Unsupported protobuf wire type: {wire_type}")


def _decode_text_index(strings: list[str], index: int) -> str:
    if index < 0 or index >= len(strings):
        return ""
    return strings[index]


def _read_packed_varints(payload: bytes) -> tuple[int, ...]:
    reader = _Reader(payload)
    values: list[int] = []
    while not reader.eof():
        values.append(reader.read_varint())
    return tuple(values)


@dataclass(slots=True)
class _RawValueType:
    type_index: int = 0
    unit_index: int = 0


def _parse_value_type(payload: bytes) -> _RawValueType:
    reader = _Reader(payload)
    result = _RawValueType()
    while not reader.eof():
        field, wire = reader.read_key()
        if wire != 0:
            reader.skip(wire)
            continue
        value = _to_signed64(reader.read_varint())
        if field == 1:
            result.type_index = value
        elif field == 2:
            result.unit_index = value
    return result


@dataclass(slots=True)
class _RawFunction:
    function_id: int = 0
    name: int = 0
    system_name: int = 0
    filename: int = 0
    start_line: int = 0


@dataclass(slots=True)
class _RawLine:
    function_id: int = 0
    line: int = 0


def _raw_lines() -> list[_RawLine]:
    return []


@dataclass(slots=True)
class _RawLocation:
    location_id: int = 0
    lines: list[_RawLine] = field(default_factory=_raw_lines)
    is_folded: bool = False


def _parse_function(payload: bytes) -> _RawFunction:
    reader = _Reader(payload)
    result = _RawFunction()
    while not reader.eof():
        field, wire = reader.read_key()
        if wire != 0:
            reader.skip(wire)
            continue
        value = reader.read_varint()
        if field == 1:
            result.function_id = value
        elif field == 2:
            result.name = value
        elif field == 3:
            result.system_name = value
        elif field == 4:
            result.filename = value
        elif field == 5:
            result.start_line = _to_signed64(value)
    return result


def _parse_line(payload: bytes) -> _RawLine:
    reader = _Reader(payload)
    result = _RawLine()
    while not reader.eof():
        field, wire = reader.read_key()
        if wire != 0:
            reader.skip(wire)
            continue
        value = reader.read_varint()
        if field == 1:
            result.function_id = value
        elif field == 2:
            result.line = _to_signed64(value)
    return result


def _parse_location(payload: bytes) -> _RawLocation:
    reader = _Reader(payload)
    result = _RawLocation()
    while not reader.eof():
        field, wire = reader.read_key()
        if field == 1 and wire == 0:
            result.location_id = reader.read_varint()
        elif field == 4 and wire == 2:
            result.lines.append(_parse_line(reader.read_length_delimited()))
        elif field == 5 and wire == 0:
            result.is_folded = bool(reader.read_varint())
        else:
            reader.skip(wire)
    return result


def _parse_sample(payload: bytes) -> Sample:
    reader = _Reader(payload)
    location_ids: list[int] = []
    values: list[int] = []
    while not reader.eof():
        field, wire = reader.read_key()
        if field == 1:
            if wire == 0:
                location_ids.append(reader.read_varint())
            elif wire == 2:
                location_ids.extend(
                    _read_packed_varints(reader.read_length_delimited())
                )
            else:
                reader.skip(wire)
        elif field == 2:
            if wire == 0:
                values.append(_to_signed64(reader.read_varint()))
            elif wire == 2:
                values.extend(
                    _to_signed64(value)
                    for value in _read_packed_varints(reader.read_length_delimited())
                )
            else:
                reader.skip(wire)
        else:
            reader.skip(wire)
    return Sample(location_ids=tuple(location_ids), values=tuple(values))


def decode_profile_bytes(data: bytes) -> Profile:
    """Decode raw or gzipped pprof protobuf bytes into the typed profile model."""
    try:
        payload = gzip.decompress(data)
    except (EOFError, zlib.error) as exc:
        raise PprofDecodeError(f"Truncated or corrupt gzip profile: {exc}") from exc
    except OSError:
        payload = data

    reader = _Reader(payload)
    strings: list[str] = []
    raw_sample_types: list[_RawValueType] = []
    raw_period_type: _RawValueType | None = None
    period = 0
    default_sample_type_index = 0
    raw_functions: list[_RawFunction] = []
    raw_locations: list[_RawLocation] = []
    samples: list[Sample] = []

    while not reader.eof():
        field, wire = reader.read_key()
        if field == 1 and wire == 2:
            raw_sample_types.append(_parse_value_type(reader.read_length_delimited()))
        elif field == 2 and wire == 2:
            samples.append(_parse_sample(reader.read_length_delimited()))
        elif field == 4 and wire == 2:
            raw_locations.append(_parse_location(reader.read_length_delimited()))
        elif field == 5 and wire == 2:
            raw_functions.append(_parse_function(reader.read_length_delimited()))
        elif field == 6 and wire == 2:
            strings.append(
                reader.read_length_delimited().decode("utf-8", errors="replace")
            )
        elif field == 11 and wire == 2:
            raw_period_type = _parse_value_type(reader.read_length_delimited())
        elif field == 12 and wire == 0:
            period = _to_signed64(reader.read_varint())
        elif field == 14 and wire == 0:
            default_sample_type_index = _to_signed64(reader.read_varint())
        else:
            reader.skip(wire)

    functions = {
        item.function_id: Function(
            function_id=item.function_id,
            name=_decode_text_index(strings, item.name),
            system_name=_decode_text_index(strings, item.system_name),
            filename=_decode_text_index(strings, item.filename),
            start_line=item.start_line,
        )
        for item in raw_functions
        if item.function_id > 0
    }
    locations = {
        item.location_id: Location(
            location_id=item.location_id,
            lines=tuple((line.function_id, line.line) for line in item.lines),
            is_folded=item.is_folded,
        )
        for item in raw_locations
        if item.location_id > 0
    }

    sample_types = tuple(
        ValueType(
            type_name=_decode_text_index(strings, raw.type_index),
            unit=_decode_text_index(strings, raw.unit_index),
        )
        for raw in raw_sample_types
    )
    period_type = (
        ValueType(
            type_name=_decode_text_index(strings, raw_period_type.type_index),
            unit=_decode_text_index(strings, raw_period_type.unit_index),
        )
        if raw_period_type is not None
        else None
    )
    default_sample_type = _decode_text_index(strings, default_sample_type_index)
    primary_value_index = select_primary_value_index(sample_types, default_sample_type)
    if primary_value_index != 0:
        samples = [
            replace(sample, primary_index=primary_value_index) for sample in samples
        ]

    return Profile(
        string_table=tuple(strings),
        functions=functions,
        locations=locations,
        samples=tuple(samples),
        sample_types=sample_types,
        period_type=period_type,
        period=period,
        default_sample_type=default_sample_type,
        primary_value_index=primary_value_index,
    )


def load_profile(path: str) -> Profile:
    with open(path, "rb") as handle:
        return decode_profile_bytes(handle.read())

File: crates/clankerprof-core/src/proto.rs
use crate::model::{select_primary_value_index, Function, Location, Profile, Sample, ValueType};
use flate2::read::GzDecoder;
use std::collections::BTreeMap;
use std::fmt::{Display, Formatter};
use std::fs;
use std::io::Read;
use std::path::Path;

#[derive(Debug)]
pub enum PprofDecodeError {
    Io(std::io::Error),
    InvalidProtobuf(String),
}

impl Display for PprofDecodeError {
    fn fmt(&self, formatter: &mut Formatter<'_>) -> std::fmt::Result {
        match self {
            Self::Io(error) => write!(formatter, "{error}"),
            Self::InvalidProtobuf(message) => formatter.write_str(message),
        }
    }
}

impl std::error::Error for PprofDecodeError {}

impl From<std::io::Error> for PprofDecodeError {
    fn from(value: std::io::Error) -> Self {
        Self::Io(value)
    }
}

type DecodeResult<T> = Result<T, PprofDecodeError>;

#[derive(Debug)]
struct Reader<'a> {
    data: &'a [u8],
    pos: usize,
}

impl<'a> Reader<'a> {
    fn new(data: &'a [u8]) -> Self {
        Self { data, pos: 0 }
    }

    fn eof(&self) -> bool {
        self.pos >= self.data.len()
    }

    fn read_byte(&mut self) -> DecodeResult<u8> {
        let Some(value) = self.data.get(self.pos).copied() else {
            return Err(PprofDecodeError::InvalidProtobuf(
                "Unexpected end of protobuf stream.".to_string(),
            ));
        };
        self.pos += 1;
        Ok(value)
    }

    fn read_varint(&mut self) -> DecodeResult<u64> {
        let mut shift = 0u32;
        let mut result = 0u64;
        loop {
            let byte = self.read_byte()?;
            // Bits past 63 are dropped, matching protobuf's 64-bit varint
            // wrap; the shift guard below caps varints at 10 bytes so this
            // can never overflow-shift.
            result |= u64::from(byte & 0x7f).wrapping_shl(shift);
            if byte < 0x80 {
                return Ok(result);
            }
            shift += 7;
            if shift >= 70 {
                return Err(PprofDecodeError::InvalidProtobuf(
                    "Invalid protobuf varint.".to_string(),
                ));
            }
        }
    }

    fn read_key(&mut self) -> DecodeResult<(u64, u8)> {
        let key = self.read_varint()?;
        Ok((key >> 3, (key & 0x07) as u8))
    }

    fn read_length_delimited(&mut self) -> DecodeResult<&'a [u8]> {
        let length = self.read_varint()? as usize;
        let end = self.pos.checked_add(length).ok_or_else(|| {
            PprofDecodeError::InvalidProtobuf(
                "Length-delimited field extends beyond stream.".to_string(),
            )
        })?;
        if end > self.data.len() {
            return Err(PprofDecodeError::InvalidProtobuf(
                "Length-delimited field extends beyond stream.".to_string(),
            ));
        }
        let payload = &self.data[self.pos..end];
        self.pos = end;
        Ok(payload)
    }

    fn skip(&mut self, wire_type: u8) -> DecodeResult<()> {
        match wire_type {
            0 => {
                self.read_varint()?;
                Ok(())
            }
            1 => self.advance(8),
            2 => {
                self.read_length_delimited()?;
                Ok(())
            }
            5 => self.advance(4),
            _ => Err(PprofDecodeError::InvalidProtobuf(format!(
                "Unsupported protobuf wire type: {wire_type}"
            ))),
        }
    }

    fn advance(&mut self, count: usize) -> DecodeResult<()> {
        let end = self.pos.checked_add(count).ok_or_else(|| {
            PprofDecodeError::InvalidProtobuf("Skip extends beyond stream.".to_string())
        })?;
        if end > self.data.len() {
            return Err(PprofDecodeError::InvalidProtobuf(
                "Skip extends beyond stream.".to_string(),
            ));
        }
        self.pos = end;
        Ok(())
    }
}

#[derive(Debug, Default)]
struct RawValueType {
    type_index: i64,
    unit_index: i64,
}

fn parse_value_type(payload: &[u8]) -> DecodeResult<RawValueType> {
    let mut reader = Reader::new(payload);
    let mut result = RawValueType::default();
    while !reader.eof() {
        let (field, wire) = reader.read_key()?;
        if wire != 0 {
            reader.skip(wire)?;
            continue;
        }
        let value = int64_from_varint(reader.read_varint()?);
        match field {
            1 => result.type_index = value,
            2 => result.unit_index = value,
            _ => {}
        }
    }
    Ok(result)
}

fn string_at_signed(strings: &[String], index: i64) -> String {
    usize::try_from(index)
        .ok()
        .and_then(|position| strings.get(position))
        .cloned()
        .unwrap_or_default()
}

#[derive(Debug, Default)]
struct RawFunction {
    function_id: u64,
    name: usize,
    system_name: usize,
    filename: usize,
    start_line: i64,
}

#[derive(Debug, Default)]
struct RawLine {
    function_id: u64,
    line: i64,
}

#[derive(Debug, Default)]
struct RawLocation {
    location_id: u64,
    lines: Vec<RawLine>,
    is_folded: bool,
}

pub fn load_profile(path: impl AsRef<Path>) -> DecodeResult<Profile> {
    decode_profile_bytes(&fs::read(path)?)
}

pub fn decode_profile_bytes(data: &[u8]) -> DecodeResult<Profile> {
    let payload = maybe_gzip_decompress(data)?;
    let mut reader = Reader::new(&payload);
    let mut strings = Vec::new();
    let mut raw_sample_types = Vec::new();
    let mut raw_period_type: Option<RawValueType> = None;
    let mut period = 0i64;
    let mut default_sample_type_index = 0i64;
    let mut raw_functions = Vec::new();
    let mut raw_locations = Vec::new();
    let mut samples = Vec::new();

    while !reader.eof() {
        let (field, wire) = reader.read_key()?;
        match (field, wire) {
            (1, 2) => raw_sample_types.push(parse_value_type(reader.read_length_delimited()?)?),
            (2, 2) => samples.push(parse_sample(reader.read_length_delimited()?)?),
            (4, 2) => raw_locations.push(parse_location(reader.read_length_delimited()?)?),
            (5, 2) => raw_functions.push(parse_function(reader.read_length_delimited()?)?),
            (6, 2) => strings.push(decode_utf8_lossy(reader.read_length_delimited()?)),
            (11, 2) => raw_period_type = Some(parse_value_type(reader.read_length_delimited()?)?),
            (12, 0) => period = int64_from_varint(reader.read_varint()?),
            (14, 0) => default_sample_type_index = int64_from_varint(reader.read_varint()?),
            _ => reader.skip(wire)?,
        }
    }

    let functions = raw_functions
        .into_iter()
        .filter(|item| item.function_id > 0)
        .map(|item| {
            (
                item.function_id,
                Function {
                    function_id: item.function_id,
                    name: string_at(&strings, item.name),
                    system_name: string_at(&strings, item.system_name),
                    filename: string_at(&strings, item.filename),
                    start_line: item.start_line,
                },
            )
        })
        .collect::<BTreeMap<_, _>>();
    let locations = raw_locations
        .into_iter()
        .filter(|item| item.location_id > 0)
        .map(|item| {
            (
                item.location_id,
                Location {
                    location_id: item.location_id,
                    lines: item
                        .lines
                        .into_iter()
                        .map(|line| (line.function_id, line.line))
                        .collect(),
                    is_folded: item.is_folded,
                },
            )
        })
        .collect::<BTreeMap<_, _>>();

    let sample_types: Vec<ValueType> = raw_sample_types
        .into_iter()
        .map(|raw| ValueType {
            type_name: string_at_signed(&strings, raw.type_index),
            unit: string_at_signed(&strings, raw.unit_index),
        })
        .collect();
    let period_type = raw_period_type.map(|raw| ValueType {
        type_name: string_at_signed(&strings, raw.type_index),
        unit: string_at_signed(&strings, raw.unit_index),
    });
    let default_sample_type = string_at_signed(&strings, default_sample_type_index);
    let primary_value_index = select_primary_value_index(&sample_types, &default_sample_type);
    for sample in &mut samples {
        sample.primary_index = primary_value_index;
    }

    Ok(Profile {
        string_table: strings,
        functions,
        locations,
        samples,
        sample_types,
        period_type,
        period,
        default_sample_type,
        primary_value_index,
    })
}

fn maybe_gzip_decompress(data: &[u8]) -> DecodeResult<Vec<u8>> {
    if data.starts_with(&[0x1f, 0x8b]) {
        let mut decoder = GzDecoder::new(data);
        let mut payload = Vec::new();
        decoder.read_to_end(&mut payload)?;
        return Ok(payload);
    }
    Ok(data.to_vec())
}

fn parse_function(payload: &[u8]) -> DecodeResult<RawFunction> {
    let mut reader = Reader::new(payload);
    let mut result = RawFunction::default();
    while !reader.eof() {
        let (field, wire) = reader.read_key()?;
        if wire != 0 {
            reader.skip(wire)?;
            continue;
        }
        let value = reader.read_varint()?;
        match field {
            1 => result.function_id = value,
            2 => result.name = value as usize,
            3 => result.system_name = value as usize,
            4 => result.filename = value as usize,
            5 => result.start_line = int64_from_varint(value),
            _ => {}
        }
    }
    Ok(result)
}

fn parse_line(payload: &[u8]) -> DecodeResult<RawLine> {
    let mut reader = Reader::new(payload);
    let mut result = RawLine::default();
    while !reader.eof() {
        let (field, wire) = reader.read_key()?;
        if wire != 0 {
            reader.skip(wire)?;
            continue;
        }
        let value = reader.read_varint()?;
        match field {
            1 => result.function_id = value,
            2 => result.line = int64_from_varint(value),
            _ => {}
        }
    }
    Ok(result)
}

fn parse_location(payload: &[u8]) -> DecodeResult<RawLocation> {
    let mut reader = Reader::new(payload);
    let mut result = RawLocation::default();
    while !reader.eof() {
        let (field, wire) = reader.read_key()?;
        match (field, wire) {
            (1, 0) => result.location_id = reader.read_varint()?,
            (4, 2) => result
                .lines
                .push(parse_line(reader.read_length_delimited()?)?),
            (5, 0) => result.is_folded = reader.read_varint()? != 0,
            _ => reader.skip(wire)?,
        }
    }
    Ok(result)
}

fn parse_sample(payload: &[u8]) -> DecodeResult<Sample> {
    let mut reader = Reader::new(payload);
    let mut location_ids = Vec::new();
    let mut values = Vec::new();
    while !reader.eof() {
        let (field, wire) = reader.read_key()?;
        match field {
            1 => match wire {
                0 => location_ids.push(reader.read_varint()?),
                2 => location_ids.extend(read_packed_varints(reader.read_length_delimited()?)?),
                _ => reader.skip(wire)?,
            },
            2 => match wire {
                0 => values.push(int64_from_varint(reader.read_varint()?)),
                2 => values.extend(
                    read_packed_varints(reader.read_length_delimited()?)?
                        .into_iter()
                        .map(int64_from_varint),
                ),
                _ => reader.skip(wire)?,
            },
            _ => reader.skip(wire)?,
        }
    }
    Ok(Sample {
        location_ids,
        values,
        primary_index: 0,
    })
}

fn read_packed_varints(payload: &[u8]) -> DecodeResult<Vec<u64>> {
    let mut reader = Reader::new(payload);
    let mut values = Vec::new();
    while !reader.eof() {
        values.push(reader.read_varint()?);
    }
    Ok(values)
}

fn decode_utf8_lossy(data: &[u8]) -> String {
    String::from_utf8_lossy(data).into_owned()
}

fn string_at(strings: &[String], index: usize) -> String {
    strings.get(index).cloned().unwrap_or_default()
}

fn int64_from_varint(value: u64) -> i64 {
    value as i64
}
